crates/ruff_linter/src/rules/flake8_pyi/rules/generic_not_last_base_class.rs

```diff
@@ -0,0 +1,139 @@
+///     ...
+/// ```
+/// ## References
+/// - [`typing.Generic` documentation](https://docs.python.org/3/library/typing.html#typing.Generic)
+///
+/// [1]: https://github.com/python/cpython/issues/106102
+#[violation]
+pub struct GenericNotLastBaseClass;
+
+impl Violation for GenericNotLastBaseClass {
+    const FIX_AVAILABILITY: FixAvailability = FixAvailability::Sometimes;
+
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("`Generic[]` should always be the last base class")
+    }
+
+    fn fix_title(&self) -> Option<String> {
+        Some("Move `Generic[]` to the end".to_string())
+    }
+}
+
+/// PYI059
+pub(crate) fn generic_not_last_base_class(checker: &mut Checker, class_def: &StmtClassDef) {
+    let bases = class_def.arguments.as_deref();
+    let Some(bases) = bases else {
+        return;
+    };
+
+    let semantic = checker.semantic();
+    if !semantic.seen_typing() {
+        return;
+    }
+
+    let generic_base_indices: Vec<usize> = bases
+        .args
+        .iter()
+        .enumerate()
+        .filter_map(|(base_index, base)| {
+            if is_generic(base, semantic) {
+                return Some(base_index);
+            }
+            None
+        })
+        .collect();
+
+    if generic_base_indices.is_empty() {
+        return;
+    }
+
+    let diagnostic =
+        {
+            if generic_base_indices.len() == 1 {
+                let base_index = generic_base_indices[0];
+                if base_index == bases.args.len() - 1 {
+                    // Don't raise issue for the last base.
+                    return;
+                }
+
+                Diagnostic::new(GenericNotLastBaseClass, class_def.identifier())
+                    .with_fix(generate_fix(bases, base_index, checker.locator()))
+            } else {
+                // No fix if multiple generics are seen in the class bases.
+                Diagnostic::new(GenericNotLastBaseClass, class_def.identifier())
+            }
+        };
+
+    checker.diagnostics.push(diagnostic);
+}
+
+/// Return `true` if the given expression resolves to `typing.Generic[...]`.
+fn is_generic(expr: &Expr, semantic: &SemanticModel) -> bool {
+    semantic.match_typing_expr(map_subscript(expr), "Generic")
+}
+
+fn generate_fix(bases: &Arguments, generic_base_index: usize, locator: &Locator) -> Fix {
+    let last_base = bases.args.last().expect("Last base should always exist");
+    let generic_base = bases
+        .args
+        .get(generic_base_index)
+        .expect("Generic base should always exist");
+    let next_base = bases
+        .args
+        .get(generic_base_index + 1)
+        .expect("Generic base should never be the last base during auto-fix");
+
+    let deletion = Edit::deletion(generic_base.start(), next_base.start());
+    let insertion = Edit::insertion(
+        format!(", {}", locator.slice(generic_base.range())),
+        last_base.end(),
+    );
+    Fix::safe_edits(insertion, [deletion])
+}
```

Comment:
what happens for pathalogical cases like

```py
class Foo(  # comment about the bracket
    # Part 1 of multiline comment 1
    # Part 2 of multiline comment 1
    Generic[T]  # comment about Generic[T]
    # another comment?
    ,  # comment about the comma?
    # part 1 of multiline comment 2
    # part 2 of multiline comment 2
    int,  # comment about int
    # yet another comment?
):  # and another one for good measure
    pass
```

Reply:
It will technically work, but comments before int and after `Generic[T]` get eaten up.

What's ruff's general stance in such cases? I know `libcst` has capabilities to try and identify which comment groups belong to which expression. But the easier thing would be to avoid moving comments at all.

Reply:
Different rules take different stances; we don't have a unified policy yet. Several of our autofixes will currently happily delete your comments (e.g. https://github.com/astral-sh/ruff/issues/9779); I think we all agree that this is bad, but we haven't yet decided _how_ bad it is (see #9790). Other rules approach comments with a perhaps-excessive level of caution, either using the raw token stream for their autofix (RUF022, RUF023), or using libcst. Both are much slower than simple text replacements, I think libcst especially.

> the fix is subjective.

absolutely! But I think that's always the case for something like this. I don't mind that much if some comments end up in the "wrong place", but I _do_ think we should try to avoid _deleting_ any comments if it's not too hard to do.

Reply:
Trying to figure out how to get the start and end ranges of the comma after the Generic base. Tried using libcst to obtain the Comma itself, but does it have position information?

Reply:
I would probably just read one character at a time after the end of the generic base until you see a comma — does that make sense here? I'd avoid LibCST unless it's absolutely necessary (it's slow).

Reply:
> I would probably just read one character at a time after the end of the generic base until you see a comma — does that make sense here? I'd avoid LibCST unless it's absolutely necessary (it's slow).

Agree -- `memchr` is probably the best tool for that (you can grep in the `crates/ruff_linter` directory for existing uses). Other possibilities are using libCST (but as Zanie says, it's really slow), or using the raw tokens (also can be slow, and probably more complicated here than using `memchr`).

If you look at https://play.ruff.rs/b2b834e8-0247-47fe-995a-c0ade0e6b240, with the following snippet:

```py
class Foo(int, str): pass
```

On the AST tab on the right, we can see:
- The `ExprName` node for the `int` base has range `10..13`
- The `ExprName` node for the `str` base has range `15..18`

On the tokens tab on the right, we can see that:
- The comma in between the two nodes has range `13..14`

Reply:
First try at this, I've used `.find()` (which I suppose can be swapped with `memchr`)  and `.position` (which I'm not sure if it can be replaced)